src/meepgeom.cpp

```diff
@@ -307,6 +308,121 @@ bool is_metal(meep::field_type ft, const material_type *material) {
     }
 }
 
+meep::vec material_grid_grad(vector3 p, material_data *md) {
+  if (!is_material_grid(md)) { meep::abort("Invalid material grid detected.\n"); }
+
+  meep::vec gradient(zero_vec(dim));
+  double *data = md->weights;
+  int nx = md->grid_size.x;
+  int ny = md->grid_size.y;
+  int nz = md->grid_size.z;
+  double rx = p.x;
+  double ry = p.y;
+  double rz = p.z;
+  int stride = 1;
+  int x1, y1, z1, x2, y2, z2;
+  double dx, dy, dz;
+
+  /* mirror boundary conditions for r just beyond the boundary */
+  rx = rx < 0.0 ? -rx : (rx > 1.0 ? 1.0 - rx : rx);
+  ry = ry < 0.0 ? -ry : (ry > 1.0 ? 1.0 - ry : ry);
+  rz = rz < 0.0 ? -rz : (rz > 1.0 ? 1.0 - rz : rz);
+
+  /* get the point corresponding to r in the epsilon array grid: */
+  x1 = meep::mirrorindex(int(rx * nx), nx);
+  y1 = meep::mirrorindex(int(ry * ny), ny);
+  z1 = meep::mirrorindex(int(rz * nz), nz);
+
+  /* get the difference between (x1,y1,z1) and the actual point */
+  dx = rx * nx - x1 - 0.5;
+  dy = ry * ny - y1 - 0.5;
+  dz = rz * nz - z1 - 0.5;
+
+  /* get the other closest point in the grid, with mirror boundaries: */
+  x2 = meep::mirrorindex(dx >= 0.0 ? x1 + 1 : x1 - 1, nx);
+  y2 = meep::mirrorindex(dy >= 0.0 ? y1 + 1 : y1 - 1, ny);
+  z2 = meep::mirrorindex(dz >= 0.0 ? z1 + 1 : z1 - 1, nz);
+
+  /* take abs(d{xyz}) to get weights for {xyz} and {xyz}2: */
+  bool signflip_dx = false, signflip_dy = false, signflip_dz = false;
+  if (dx != fabs(dx)) {
+    dx = fabs(dx);
+    signflip_dx = true;
+  }
+  if (dy != fabs(dy)) {
+    dy = fabs(dy);
+    signflip_dy = true;
+  }
+  if (dz != fabs(dz)) {
+    dz = fabs(dz);
+    signflip_dz = true;
+  }
```

Comment:
Can we also share some code here with `linear_interpolate`?

Reply:
Fixed.